### PR TITLE
Optional Dependencies

### PR DESCRIPTION
**Pull request by askvortsov1**
**Refs https://github.com/flarum/core/issues/1318**

**Changes proposed in this pull request:**
- Add "optional dependencies" as per #1318
- Reintroduce extension dependency resolution from https://github.com/flarum/core/pull/2317
- Sort extensions every time an extension is enabled/disabled

**Reviewers should focus on:**
- I have mixed feelings about the name "optional dependencies". Maybe something like "bootBeforeTheseExtensions" would be better? But that's very verbose and would make for a clunky getter method name.

**Confirmed**

~~- [ ] Frontend changes: tested on a local Flarum installation.~~
- [x] Backend changes: tests are green (run `composer test`).

**Required changes:**

- [ ] Related documentation PR: (Remove if irrelevant)
- [ ] Related core extension PRs: (Remove if irrelevant)

### CODE REVIEW

File: src/Extension/Extension.php [hunk 1]
```diff
@@ -203,16 +211,30 @@ public function setVersion($version)
...
      *                             are flarum extensions.
+     * @param array $enabledIds:   An associative array where keys are the composer package names
+     *                             of enabled extensions. Used to figure out optional dependencies.
      */
-    public function calculateDependencies($extensionSet)
+    public function calculateDependencies($extensionSet, $enabledIds)
     {
         $this->extensionDependencyIds = (new Collection(Arr::get($this->composerJson, 'require', [])))
             ->keys()
+            ->merge(Arr::get($this->composerJson, 'extra.flarum-extension.optional-dependencies', []))
```
Comment (KyrneDev) [inline src/Extension/Extension.php:221]: What's the purpose of including the optional dependencies with the "hard" dependencies? If we already get the optional ones later, then merge the two in the order function, wouldn't it make sense to keep them separate?

Comment (askvortsov1) [reply src/Extension/Extension.php]: Ah, yeah that was a mistake

File: src/Extension/Extension.php [hunk 2]
```diff
@@ -203,16 +211,30 @@ public function setVersion($version)
...
-            })->map(function ($key) {
+            })
+            ->map(function ($key) {
+                return static::nameToId($key);
+            })
+            ->toArray();
+
+        $this->optionalDependencyIds = (new Collection(Arr::get($this->composerJson, 'extra.flarum-extension.optional-dependencies', [])))
+            ->map(function ($key) {
                 return static::nameToId($key);
```
Comment (askvortsov1) [inline src/Extension/Extension.php:231]: Should optional dependencies be provided as composer package names or Flarum extension IDs? I'm leaning towards switching it to the latter since this is Flarum-specific config

Comment (datitisev) [reply src/Extension/Extension.php:231]: Hm... multiple extensions can technically resolve to the same ID, so I feel like we'd want dependencies to use the package name. Specially as that's used for required dependencies (ignoring the fact that uses composer require in composer.json)

Comment (askvortsov1) [reply src/Extension/Extension.php:231]: Thing is, the composer package name isn't actually used anywhere in calculations; in the lines directly below here, it turns the composer package names into flarum extension IDs anyway.

Although on the other hand, if we ever decide to switch back to using pure composer package names, having that be the provided format now will reduce breaks then.

Comment (SychO9) [reply src/Extension/Extension.php:231]: I would lean towards using package names, just so that we're using one single format when referring to extensions in composer.json (consistency).

Comment (askvortsov1) [reply src/Extension/Extension.php:231]: In that case, let's keep as is

File: src/Extension/ExtensionManager.php [hunk 3]
```diff
@@ -86,7 +86,9 @@ public function getExtensions()
             // We calculate and store a set of composer package names for all installed Flarum extensions,
             // so we know what is and isn't a flarum extension in `calculateDependencies`.
             // Using keys of an associative array allows us to do these checks in constant time.
+            // We do the same for enabled extensions, for optional dependencies.
             $installedSet = [];
+            $enabledIds = array_flip($this->getEnabled());
```
Comment (SychO9) [inline src/Extension/ExtensionManager.php:91]: This is probably just a nitpick, but leaving it unflipped and then using `in_array` instead of `array_key_exists` seems more simple

**Review by SychO9 — COMMENTED** [review]
Code looks good!

File: src/Extension/ExtensionManager.php (hunk 3 above)
Comment (askvortsov1) [reply src/Extension/ExtensionManager.php:91]: The rationale here was to use constant time key-based lookup instead of linear time value, since multiple extensions will perform multiple lookups.

Comment (SychO9) [reply src/Extension/ExtensionManager.php:91]: I see, I hadn't noticed or thought about it before, that's better performance wise, and since we're certain that the values are unique than it works just fine, nice.

Comment (askvortsov1) [reply src/Extension/ExtensionManager.php:91]: Any reason for not explicitly approving the PR?

Comment (SychO9) [reply src/Extension/ExtensionManager.php:91]: running some tests locally right now

File: src/Extension/ExtensionManager.php [hunk 4]
```diff
@@ -382,4 +392,86 @@ public static function pluckTitles(array $exts)
...
+        $extensionGraph = [];
+        $output = [];
+        $missingDependencies = []; // Extensions are invalid if they are missing dependencies, or have circular dependencies.
+        $circularDependencies = [];
+        $pendingQueue = [];
+        $inDegreeCount = []; // How many extensions are dependent on a given extension?
+
+        foreach ($extensionList as $extension) {
+            $extensionIdMapping[$extension->getId()] = $extension;
+            $extensionGraph[$extension->getId()] = array_merge($extension->getExtensionDependencyIds(), $extension->getOptionalDependencyIds());
```
Comment (SychO9) [inline src/Extension/ExtensionManager.php:422]: could it be because we're merging the required extensions array with the optional extensions array ?

**Review by SychO9 — CHANGES_REQUESTED** [review]
So I noticed an odd behaviour,

I created two dummy extensions that have `flarum/tags` as an optional dependency, I enabled both of them and observed the enabled extensions list order, looked fine and logical. I then proceeded to disable the tags extension, being an optional dependency it should not affect the other two extensions, however everytime I tried doing so one of the extensions (not always the same one) gets disabled.

Tracked it down to an issue in the extension order resolution algorithm.

File: src/Extension/ExtensionManager.php (hunk 4 above)
Comment (askvortsov1) [reply src/Extension/ExtensionManager.php:422]: Ah, good thing you tested! That's an edge case I didn't think of (and that needs tests). Merging is correct, but we need to filter the optional extensions array to only have installed extensions.

Comment (askvortsov1) [reply src/Extension/ExtensionManager.php]: Added a test, made sure that test failed with the current code, implemented the solution. Test now passes. Thanks for finding this!

**Review by SychO9 — APPROVED** [review]

**Review by KyrneDev — APPROVED** [review]
Can confirm the bug reported by @SychO9 is fixed, works well locally!
